storage-retrieval/lib/src/client.rs

```diff
@@ -0,0 +1,252 @@
+            .for_each(|root| self.request_block(root));
+
+        // Timeout future
+        let mut timeout: BoxFuture<()> = self
+            .timeout
+            .map(|t| sleep(t).boxed())
+            .unwrap_or_else(|| std::future::pending().boxed());
+
+        loop {
+            tokio::select! {
+                // Data download timeout
+                _ = &mut timeout => {
+                    // Return an error as indication that the download timed out
+                    return Err(ClientError::DownloadTimeout);
+                }
+                // Handle events received when we get some blocks back
+                event = self.swarm.select_next_some() => {
+                    // Handle event received from the providers
+                    self.on_swarm_event(event).await?;
+
+                    // if no inflight queries, that means we received
+                    // everything requested. Finalize the blockstore.
+                    if self.queries.is_empty() {
+                        self.blockstore.finalize(None).await?;
+                        break;
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn request_block(&mut self, cid: Cid) {
+        debug!("requesting block {cid}");
+        let query_id = self.swarm.behaviour_mut().bitswap.get(&cid);
+        self.queries.insert(query_id, cid);
+    }
+
+    async fn on_swarm_event(
+        &mut self,
+        event: SwarmEvent<BehaviourEvent<EmptyBlockstore>>,
+    ) -> Result<(), ClientError> {
+        trace!(?event, "Received swarm event");
+
+        match event {
+            SwarmEvent::ConnectionEstablished {
+                peer_id,
+                connection_id,
+                endpoint,
+                ..
+            } => {
+                self.on_peer_connected(peer_id, connection_id, endpoint);
+            }
+            SwarmEvent::ConnectionClosed {
+                peer_id,
+                connection_id,
+                ..
+            } => {
+                self.on_peer_disconnected(peer_id, connection_id);
+            }
+            SwarmEvent::Behaviour(BehaviourEvent::Bitswap(event)) => {
+                self.on_bitswap_event(event).await?;
+            }
+            _ => {
+                // Nothing to do here
+            }
+        }
+
+        Ok(())
+    }
+
+    #[instrument(skip_all, fields(peer_id = %peer_id))]
+    fn on_peer_connected(
+        &mut self,
+        peer_id: PeerId,
+        _connection_id: ConnectionId,
+        _endpoint: ConnectedPoint,
+    ) {
+        debug!("Peer connected");
+    }
+
+    #[instrument(skip_all, fields(peer_id = %peer_id))]
+    fn on_peer_disconnected(&mut self, peer_id: PeerId, _connection_id: ConnectionId) {
+        debug!("Peer disconnected");
+    }
+
+    #[instrument(level = "trace", skip(self))]
+    async fn on_bitswap_event(&mut self, event: beetswap::Event) -> Result<(), ClientError> {
+        match event {
```

Comment:
This would shorten the following code
```suggestion
        use beetswap::Event;
        match event {
```